### PR TITLE
planner, expression: pushdown AggFuncMode to coprocessor

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/30923
ref: 
https://github.com/pingcap/tipb/pull/254
https://github.com/pingcap/tics/pull/3822

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: expression/aggregation/descriptor.go [hunk 1]
```diff
@@ -110,6 +110,7 @@ func (a *AggFuncDesc) Clone() *AggFuncDesc {
 	for i, byItem := range a.OrderByItems {
 		clone.OrderByItems[i] = byItem.Clone()
 	}
+	clone.Mode = a.Mode
```
Comment (fzhedu) [inline expression/aggregation/descriptor.go:113]: dose previous function like Equal() and window functions need consider Agg mode?

File: expression/aggregation/agg_to_pb.go [hunk 2]
```diff
@@ -102,9 +102,47 @@ func AggFuncToPBExpr(sctx sessionctx.Context, client kv.Client, aggFunc *AggFunc
...
+		*pbMode = tipb.AggFunctionMode_Partial2Mode
+	case DedupMode:
+		*pbMode = tipb.AggFunctionMode_DedupMode
+	}
+	return pbMode
+}
+
+func pbAggFunctionModeToAggFunctionMode(pbMode *tipb.AggFunctionMode) (mode AggFunctionMode) {
+	// set the default mode to Partial1Mode since before adding aggFunctionMode to tipb,
+	// the aggregation mode in is always Partial1Mode
```
Comment (fzhedu) [inline expression/aggregation/agg_to_pb.go:129]: func PBAggFun...?
what is the comments meaning?

File: planner/core/task.go [hunk 3]
```diff
@@ -1967,7 +1976,7 @@ func RemoveUnnecessaryFirstRow(
 				}
 				if gbyExpr.Equal(sctx, aggFunc.Args[0]) {
 					canOptimize = true
-					funcMap[aggFunc].Args[0] = finalGbyItems[j]
+					firstRowFuncMap[aggFunc].Args[0] = finalGbyItems[j]
```
Comment (fzhedu) [inline planner/core/task.go:1979]: `gbyExpr.Equal()` is judged by the number semantic？if yes, '00' = 'gg'?

Comment (XuHuaiyu) [reply planner/core/task.go:1979]: The `Equal` function is an attribute method of the Expression interface.
It seems not judged by the number semantic.

BTW, it is not modified by this commit.

**Review by guo-shaoge — APPROVED** [review]

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**